File: pkg/tmpl/tmpl.go
// Copyright (c) 2026 Adam Ousmer. All rights reserved.
// Proprietary. No license granted. See LICENSE.md.

package tmpl

import "strings"

type Kind uint8

const (
	KindLiteral Kind = iota
	KindVar
)

type Token struct {
	Kind Kind

	Text string

	Name        string
	Payload     string
	HasPayload  bool
	Fallback    string
	HasFallback bool
	Raw         string
}

func (t Token) Key() string {
	if !t.HasPayload {
		return t.Name
	}
	return t.Name + ":" + t.Payload
}

func (t Token) Resolve(val string, ok bool) string {
	switch {
	case !ok:
		return t.Raw
	case val == "":
		return t.Fallback
	default:
		return val
	}
}

func Lex(s string) []Token {
	var out []Token
	from := 0
	for i := 0; i < len(s); {
		if s[i] != '{' {
			i++
			continue
		}
		end := closeBrace(s, i+1)
		if end < 0 {
			break
		}
		out = appendLiteral(out, s[from:i])
		out = append(out, parseSpan(s[i:end+1]))
		i = end + 1
		from = i
	}
	return appendLiteral(out, s[from:])
}

func appendLiteral(out []Token, text string) []Token {
	if text == "" {
		return out
	}
	return append(out, Token{Kind: KindLiteral, Text: text})
}

func Append(dst []byte, s string, repl func(tok Token) (val string, ok bool)) []byte {
	for i := 0; i < len(s); {
		if s[i] != '{' {
			dst = append(dst, s[i])
			i++
			continue
		}
		end := closeBrace(s, i+1)
		if end < 0 {
			return append(dst, s[i:]...)
		}
		dst = appendSpan(dst, s[i:end+1], repl)
		i = end + 1
	}
	return dst
}

func Expand(s string, repl func(tok Token) (val string, ok bool)) string {
	if s == "" {
		return ""
	}
	return string(Append(make([]byte, 0, len(s)+32), s, repl))
}

func closeBrace(s string, from int) int {
	for j := from; j < len(s); j++ {
		if s[j] == '}' {
			return j
		}
	}
	return -1
}

func appendSpan(dst []byte, raw string, repl func(tok Token) (val string, ok bool)) []byte {
	tok := parseSpan(raw)
	cond, isCond := tok.Cond()
	if !isCond {
		return append(dst, tok.Resolve(repl(tok))...)
	}
	val, known := repl(cond.Ref)
	return append(dst, tok.CondText(cond, val, known)...)
}

func parseSpan(raw string) Token {
	body := raw[1 : len(raw)-1]
	tok := Token{Kind: KindVar, Raw: raw}
	if i := strings.LastIndexByte(body, '|'); i >= 0 {
		tok.Fallback, tok.HasFallback = body[i+1:], true
		body = body[:i]
	}
	name, payload, found := strings.Cut(body, ":")
	tok.Name, tok.Payload, tok.HasPayload = strings.ToLower(name), payload, found
	return tok
}

File: web/kit/lib/engine/tmpl.ts
// Copyright (c) 2026 Adam Ousmer. All rights reserved.
// Proprietary. No license granted. See LICENSE.md.

export interface LiteralToken {
  kind: 'literal';
  text: string;
}

export interface VarToken {
  kind: 'var';
  name: string;
  payload: string | null;
  fallback: string | null;
  raw: string;
  key: string;
}

export type Token = LiteralToken | VarToken;

export function lex(template: string): Token[] {
  const out: Token[] = [];
  let from = 0;
  let i = 0;
  while (i < template.length) {
    if (template[i] !== '{') {
      i++;
      continue;
    }
    const end = template.indexOf('}', i + 1);
    if (end < 0) break;
    pushLiteral(out, template.slice(from, i));
    out.push(parseSpan(template.slice(i, end + 1)));
    i = end + 1;
    from = i;
  }
  pushLiteral(out, template.slice(from));
  return out;
}

function pushLiteral(out: Token[], text: string): void {
  if (text !== '') out.push({ kind: 'literal', text });
}

function parseSpan(raw: string): VarToken {
  let body = raw.slice(1, -1);
  let fallback: string | null = null;
  const pipe = body.lastIndexOf('|');
  if (pipe >= 0) {
    fallback = body.slice(pipe + 1);
    body = body.slice(0, pipe);
  }
  const colon = body.indexOf(':');
  if (colon < 0) {
    const name = body.toLowerCase();
    return { kind: 'var', name, payload: null, fallback, raw, key: name };
  }
  const name = body.slice(0, colon).toLowerCase();
  const payload = body.slice(colon + 1);
  return { kind: 'var', name, payload, fallback, raw, key: `${name}:${payload}` };
}

export function resolveToken(token: VarToken, value: string | null): string {
  if (value === null) return token.raw;
  if (value === '') return token.fallback ?? '';
  return value;
}

export function intactSpan(name: string, payload: string | null): string | null {
  const span = payload === null ? `{${name}}` : `{${name}:${payload}}`;
  const tokens = lex(span);
  if (tokens.length !== 1) return null;
  const token = tokens[0];
  if (token.kind !== 'var') return null;
  if (token.name !== name || token.payload !== payload) return null;
  return token.fallback === null ? span : null;
}

export function expand(template: string, resolve: (token: VarToken) => string | null): string {
  return lex(template)
    .map((token) => (token.kind === 'literal' ? token.text : renderSpan(token, resolve)))
    .join('');
}

function renderSpan(token: VarToken, resolve: (token: VarToken) => string | null): string {
  const cond = parseCond(token);
  if (cond === null) return resolveToken(token, resolve(token));
  return condText(token, cond, resolve(cond.ref));
}

const COND_NAME = 'if';

export interface Cond {
  ref: VarToken;
  want: string | null;
  then: string;
  els: string;
}

export function parseCond(token: VarToken): Cond | null {
  if (token.name !== COND_NAME || token.payload === null) return null;
  const parts = token.payload.split(':');
  const last = parts.length - 1;
  if (last < 1) return null;
  const key = last === 1 ? parts[0] : parts.slice(0, last - 1).join(':');
  const then = last === 1 ? parts[1] : parts[last - 1];
  const els = last === 1 ? '' : parts[last];
  const eq = key.indexOf('=');
  if (eq < 0) return { ref: refToken(key), want: null, then, els };
  return { ref: refToken(key.slice(0, eq)), want: key.slice(eq + 1), then, els };
}

function refToken(key: string): VarToken {
  const colon = key.indexOf(':');
  if (colon < 0) {
    const name = key.toLowerCase();
    return { kind: 'var', name, payload: null, fallback: null, raw: '', key: name };
  }
  const name = key.slice(0, colon).toLowerCase();
  const payload = key.slice(colon + 1);
  return { kind: 'var', name, payload, fallback: null, raw: '', key: `${name}:${payload}` };
}

export function condHolds(cond: Cond, value: string): boolean {
  return cond.want === null ? value !== '' : value === cond.want;
}

export function condText(token: VarToken, cond: Cond, value: string | null): string {
  if (value === null) return token.raw;
  return condHolds(cond, value) ? cond.then : cond.els;
}
